README.md
regbits: C++ templates for type-safe bit manipulation
====================================================

**regbits** is a C++ template library for type-safe manipulation of bits in microcontroller peripheral registers. Compiled executable object code equals or outperforms traditional techniques in terms of size and speed, and complexity of header files describing peripheral/register layout is comparable.



Contents  <a name="contents"></a>
--------
* [License](#license)
* [What does regbits do?](#what_does_regbits_do)
* [What is regbits for?](#what_is_regbits_for)
* [How is regbits used?](#how_is_regbits_used)
* [How is regbits implemented?](#how_is_regbits_implemented?)
* [Comparison with other approaches](#comparison_with_other_approaches)
    * [Raw numeric values](#raw_numeric_values)
    * [Preprocessor symbolic peripheral addresses and register offsets](#preprocessor_symbolic_peripheral_addresses_and_register_offsets)
    * [Preprocessor symbolic bit definitions](#preprocessor_symbolic_bit_definitions)
    * [Static consts](#static_consts)
    * [C/C++ enums](#c_cxx_enums)
    * [C structs](#c_structs)
    * [C bitfields](#c_bitfields)
    * [Helper functions](#helper_functions)
    * [Higher-level wrappers](#Higher-level_wrappers)
* [Requirements, limitations, problems](#requirements_limitations_problems)
* [Unit test](#unit_test)
    * [Type-safe error detection](#type_safe_error_detection)
    * [Computational correctness](#computational_correctness)
    * [Performance claims](#performance_claims)
* [Future enhancements](#future_enhancements)


License  <a name="license"></a>
-------

regbits: C++ templates for type-safe bit manipulation

Copyright (C) 2019 Mark R. Rubin

This file is part of regbits.

The regbits program is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, either version 3 of the License, or (at your option) any later version.

The regbits program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more details.

You should have received a copy of the [GNU General Public License](LICENSE.txt) along with the regbits program.  If not, see <https://www.gnu.org/licenses/gpl.html>


What does regbits do? <a name="what_does_regbits_do"></a>
-------------------------

Regbits enables easy to use, optimally efficient low-level access to [microcontroller registers](#microcontroller_registers).

Given appropriate header files, the following all compile to executable object code of equivalent size and speed (see [Performance claims](#performance_claims), below):

        // raw address and bits
        *(int*)(0x40008008) = 0xc0;

        // symbolic names for peripheral, register, and bits
        *(TIMER1 + TIMER_CTRL_OFFSET) = TIMER_ENABLE | TIMER_COUNT_UP;

        // C struct peripheral definition with symbolic bit names
        TIMER1->CTRL = TIMER_ENABLE | TIMER_COUNT_UP;

        // C++ regbits
        timer1->ctrl = TIMER_ENABLE | TIMER_COUNT_UP;

Things get more interesting when manipulating contiguous spans of several bits:

        // raw address and bits
        *(int*)(0x400c0004) = (*(int*)(0x400c0004) & ~0xf00) | 0x600;

        // symbolic names for peripheral, register, and bits
        *(TIMER3 + TIMER_CONFIG_OFFSET) = (*(TIMER3 + TIMER_CONFIG_OFFSET) & TIMER_CLOCK_DIV_MASK) | TIMER_CLOCK_DIV_8;

        // C struct peripheral definition with symbolic bit names and bit manipulation preprocessor macro
        INSERT_BITS(USART2->CONFIG, USART_BAUD_MASK, USART_BAUD_19200);

        // C++ regbits
        usart2->config /= USART_BAUD_19200;  // or usart2->config.ins(USART_BAUD_19200)

But more important than what regbits does is what it does NOT do. All of the following will compile without error but almost certainly cause serious, hard to debug problems when executed:

        // read the manual if you want to know what this is trying to do
        *(int*)(0x40100010) = 0x802;

        // they probably use the same bit positions for all peripherals
        *(SPI0 + SPI_CONFIG_OFFSET) = I2C_ENABLE | I2C_MASTER_MODE;

        // this might do something interesting
        ADC1->CTRL = USART_PARITY_EVEN | TIMER_ENABLE_INTERRUPT;
        
        // we'll catch any problems during debug and test
        init_usb_periph(USART3);

        // hope this works when used later on
        uint32_t timer_config_settings = SPI_LSB_FIRST | I2C_NACK_ENABLE;

Compared to regbits which won't allow any of the following to compile: <a name="regbits_compile_time_errors"></a>

        // ERROR: can't set SPI peripheral with I2C configuration constants
        spi0->ctrl = I2C_ENABLE;

        // ERROR: can't set register with another register's constant
        i2c1->interrupts = I2C_MODE_FAST;

        // ERROR: can't call function with wrong kind of peripheral
        init_usb(USART3);
        
        // ERROR: can't mix different peripherals' constants
        uint32_t timer_config_settings = (TIMER_COUNT_DOWN | USART_BAUD_9600).bits();

These are contrived, exaggerated examples. However, the point remains that this kind of error is easy to make, hard to detect by source code inspection, and possibly even harder to find during runtime testing and debugging. A one-letter typo can easily hide in thousands of lines of code:

        I2C3->CTRL = I2C_TIMEOUT_ON | I2S_ENABLE | I2C_MASTER_MODE;



What is regbits for? <a name="what_is_regbits_for"></a>
--------------------

Microcontrollers (MCUs / "Systems On A Chip" / SOCs) typically include *peripherals*: Hardware functionality such as input/output ports, timers, serial communications, analog-to-digital and digital-to-analog conversion, etc.

<a name="microcontroller_registers"></a>
These peripherals typically contain multiple *registers*: Sets of memory locations (usually contiguous) which when written to and read from control the peripheral and access its data. Additionally, the MCU itself -- clock speed, interrupts, I/O configuration -- is configured via similar sets of registers.

Single and spans of multiple (usually contiguous) bits within each register typically control or access specific capabilities of the peripheral or MCU. A register for a serial communications peripheral might be structured:

          31  30  29  28  27  26  25  24  23  22  21  20  19  18  17  16  15  14  13  12  11  10  9   8   7   6   5   4   3   2   1   0
        |   |   |   |   |   |   |   | P |   |   |       PORT        |   | TR| RR|  CLK  |   |   | TI| RI|   |   |   |   SPEED   | M | E |

        bit(s)  label   functionality   
          0       E     Enable peripheral
          1       M     Mode (0==transmit, 1==receive)
         2-4    SPEED   Speed (0==slowest, 7==fastest)
          8      RI     Enable receive interrupt
          9      TI     Enable transmit interrupt
        12-13    CLK    Clock source (0==peripherals, 1==main, 2==external)
         14      RR     Receive data available
         15      TR     Transmit ready
        17-21   PORT    I/O port
         24      P      Polarity (0=positive, 1=negative)

Although the examples in this `README.md` and the [`unittest`](unittest) code are for 32-bit architectures, all are equally applicable to 64-bit and other word widths. And although designed for register access, regbits and similar methods are not limited to such usage.


#### Hardware/software limitations

All or almost all contemporary MCU hardware permits accessing memory by words (32/64/etc bits), bytes (8 bits), shorts (typically 16 bits) and other fixed, multiples-of-8, bit widths (32 bits on 64-bit architecture, etc).

Software frequently needs to access register bits (single or multiple) as separate, isolated atomic units, changing only those bits without affecting others in the same register.  Hardware does not support this, and with the exception of C bitfields ([see below](#c_bitfields)), neither do the C and C++ languages. Regbits and other software techniques (see [below](#comparison_with_other_approaches)) address these limitations.


#### Register memory accesses <a name="register_memory_access"></a>

Although registers are accessed as memory locations, they may or may not behave as such. C and C++ declarations should always include the `volatile` keyword because register bits may change independent of the CPU activity and the C/C++ execution model. There can be further differences; see [Future enhancements](#future_enhancements), below.



How is regbits used? <a name="how_is_regbits_used"></a>
--------------------

Given a microcontroller definition file based on the regbits C++ class templates (see [below](#how_is_regbits_implemented?)), the set of operators and functions defined in [`regbits.hxx`](regbits.hxx) can be used in client code.

The examples below are based on the sample definition file included in this repository's unit test directory, [`mcu_regbits.hxx`](unittest/mcu_regbits.hxx) (see [Unit test](#unit_test), below). This file contains a small set of example peripheral and register definitions describing a fictitious MCU and some of its peripherals.

The examples show correct regbits usage. The type-safe design of regbits will intentionally produce compilation failures in case of mismatches between regbits objects, as per [above](#regbits_compile_time_errors). Also see ["Type-safe error detection"](#type_safe_error_detection) in the unittest documentation, below.

Many of the examples (and the [unit test suite](unittest)) show regbits' operator and function/method syntaxes, both of which produce the same results and can be used interchangeably.


#### Single bits vs bit spans

Regbits defines two C++ classes for manipulating regisers: `Bits` and `Mskd`.

`Bits` are used for setting/clearing/testing/etc single bits. The `Bits` (from [`regbits.hxx`](regbits.hxx) used in the following examples are:

        Serial::Config::POLARITY   // bit: 1<<6
        Serial::Config::ENDIAN     // bit: 1<<5
        Serial::Config::MODE       // bit: 1<<0

`Mskd` objects represent a (usually) contiguous bit span and contain a bit mask:

        Serial::Config::DATALEN_16_BITS         // bits:  1<<1   mask: 0x03<<1
        Timer::Prescale::PRESCALER_HIGH<17>()   // bits: 17<<25  mask: 0x1f<<25

The bit mask allows setting/testing/etc bit spans to the exact "bits" pattern, both setting and clearing bits within the mask, as necessary.

Single `Bits` and `Mskd` bit spans can largely be used interchangeably, with a few exceptions. `Bits` and `Mskd` objects can be combined with the `|` operator in either order, but the type of the result is different:
        
        Serial::Config::DATALEN_16_BITS | Serial::Config::MODE   // Mskd|Bits returns Mskd
        Serial::Config::MODE | Serial::Config::DATALEN_16_BITS   // Bits|Mskd returns Bits, losing mask information

The `Mskd` type's mask information is important for operations such as `/=`, `ins()`, `all()`, `>`, `>=`, `<`, and `<=`, described below. `Bits` objects will be automatically be converted to `Mskd` as required, but with a trivial mask equal to the bits themselves; thus losing a `Mskd`'s mask via the `|` operator may produce unexpected results. In practice these issues are largely intuitive and obvious. Note, however, that when masking information is not needed, `Bits` are more efficient than `Mskd` spans.


#### Set register to single bit, bit span, or numeric value

        // set register to single bit
        serial2->config =   Serial::Config::POLARITY ;
        serial2->config.wrt(Serial::Config::POLARITY);

        // set register to multiple single bits
        serial2->config =   Serial::Config::ENDIAN | Serial::Config::POLARITY ;
        serial2->config.wrt(Serial::Config::ENDIAN | Serial::Config::POLARITY);

        // set register to bit span
        serial2->config =   Serial::Config::DATALEN_16_BITS ;  // datalen 8, 16, or 32 encoded in 2 bits
        serial2->config.wrt(Serial::Config::DATALEN_16_BITS);

        // set register to multiple bit spans
        serial2->config =   Serial::Config::DATALEN_16_BITS | Serial::Config::PARITY_EVEN ;  // parity EVEN, ODD, CRC encoded in 2 bits
        serial2->config.wrt(Serial::Config::DATALEN_16_BITS | Serial::Config::PARITY_EVEN);

        // set register to mix of single and bit span
        serial2->config =   Serial::Config::DATALEN_16_BITS | Serial::Config::MODE ;  // mode is single bit
        serial2->config.wrt(Serial::Config::DATALEN_16_BITS | Serial::Config::MODE);
        // order unimportant
        serial2->config =   Serial::Config::MODE | Serial::Config::DATALEN_16_BITS ;
        serial2->config.wrt(Serial::Config::MODE | Serial::Config::DATALEN_16_BITS);

        // set register to numeric value
        serial2->config =   0x12345678 ;
        serial2->config.wrt(0x12345678);


#### Set, clear, flip bits

        // set single/multiple bits/spans and mixtures thereof
        serial2->config |=  Serial::Config::POLARITY;
        serial2->config.set(Serial::Config::POLARITY        | Serial::Config::ENDIAN     );
        serial2->config |=  Serial::Config::DATALEN_16_BITS | Serial::Config::PARITY_EVEN ;
        serial2->config.set(Serial::Config::POLARITY        | Serial::Config::ENDIAN     );


        // clr single/multiple bits/spans and mixtures thereof
        serial2->config -=  Serial::Config::POLARITY;
        serial2->config.clr(Serial::Config::POLARITY        | Serial::Config::ENDIAN     );

RELEASE_NOTES.txt
1.1.0 (regbits)  0.1.4 (svd2regbits.py)	  2021 Feb 02
-----------------------------------------------------
* svd2regbits.py: Support multiple "<interrupt>" tags in "<peripheral>"
* regbits.hxx: (no changes)

1.1.0 (regbits)  0.1.3 (svd2regbits.py)	  2021 Jan 13
-----------------------------------------------------
* svd2regbits.py:
  * Removed nested arm namespace in .hxx output
  * Fix <enumeratedValues> for <register><fields><field derivedFrom=...>
* regbits.hxx: (no changes)

1.1.0 (regbits)  0.1.2 (svd2regbits.py)	  2021 Jan 07
-----------------------------------------------------
* svd2regbits.py: Handle "<dimIncrement>" for clusters
* regbits.hxx: (no changes)

1.1.0 (regbits)  0.1.1 (svd2regbits.py)	  2020 Dec 27
-----------------------------------------------------
* svd2regbits.py: Handle "<enumeratedValues derivedFrom=...>"
* regbits.hxx: (no changes)

1.1.0	2020 Dec 23
-------------------
* Added REGBITS_MSKD_UNSGN for WORD sizes wider than unsigned
* Small fix to REGBITS_ARRAY_RANGE constexpr static_assert
* New svd2regbits.py utility, and README_svd2regbits.md

1.0.2	2020 Nov 10
-------------------
* Minor change in regbits.hxx (Reg::zero constexpr)
* Major rewrite of unittest build system:
  - top-level Makefile with make/cleanbin/clean targets
  - bad/{arm,linux} makefiles build both pass and intentional fail tests
  - rework capture of arm subdirs compilation results

1.0.1	2020 Jan 19
-------------------
* Reg extractors return Bits or Mskd instead of Reg
* Added bool Reg::all(Bits mask, Bits bits)
* Added "explicit" keyword to additional constructors
* "= default" copy constructors (required for GCC 9)
* Version numbering


unittest/mcu_raw.h
// regbits: C++ templates for type-safe bit manipulation
// Copyright (C) 2019 Mark R. Rubin
//
// This file is part of regbits.
//
// The regbits program is free software: you can redistribute it and/or
// modify it under the terms of the GNU General Public License as
// published by the Free Software Foundation, either version 3 of the
// License, or (at your option) any later version.
//
// The regbits program is distributed in the hope that it will be
// useful, but WITHOUT ANY WARRANTY; without even the implied warranty
// of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
// General Public License for more details.
//
// You should have received a copy of the GNU General Public
// License (LICENSE.txt) along with the regbits program.  If not, see
// <https://www.gnu.org/licenses/gpl.html>


#ifndef MCU_RAW_H
#define MCU_RAW_H

#include <stdint.h>



#ifndef REGBITS_PERIPH_BASE
#error Must define REGBITS_PERIPH_BASE via -DPERIPH_BASE=xxxx on commandline
#endif

#define GPIO_BASE       ((REGBITS_PERIPH_BASE) + 0x00000000)
#define TIMER_BASE      ((REGBITS_PERIPH_BASE) + 0x00000200)
#define SERIAL_BASE     ((REGBITS_PERIPH_BASE) + 0x00000280)

#define GPIO0_BASE      ((GPIO_BASE)   + 0x00000000)
#define GPIO1_BASE      ((GPIO_BASE)   + 0x00000100)

#define TIMER0_BASE     ((TIMER_BASE)  + 0x00000000)
#define TIMER1_BASE     ((TIMER_BASE)  + 0x00000020)
#define TIMER2_BASE     ((TIMER_BASE)  + 0x00000040)
#define TIMER3_BASE     ((TIMER_BASE)  + 0x00000060)

#define SERIAL0_BASE    ((SERIAL_BASE) + 0x00000000)
#define SERIAL1_BASE    ((SERIAL_BASE) + 0x00000020)
#define SERIAL2_BASE    ((SERIAL_BASE) + 0x00000040)

#define GPIO0           ((volatile uint32_t*)(GPIO0_BASE))
#define GPIO1           ((volatile uint32_t*)(GPIO1_BASE))

#define TIMER0          ((volatile uint32_t*)(TIMER0_BASE))
#define TIMER1          ((volatile uint32_t*)(TIMER1_BASE))
#define TIMER2          ((volatile uint32_t*)(TIMER2_BASE))
#define TIMER3          ((volatile uint32_t*)(TIMER3_BASE))

#define SERIAL0         ((volatile uint32_t*)(SERIAL0_BASE))
#define SERIAL1         ((volatile uint32_t*)(SERIAL1_BASE))
#define SERIAL2         ((volatile uint32_t*)(SERIAL2_BASE))

#define GPIO_BYTES_OFFSET           0
#define GPIO_WORDS_OFFSET           8
#define GPIO_SET_OFFSET            40
#define GPIO_CLR_OFFSET            41

#define TIMER_CONTROL_OFFSET        0
#define TIMER_PRESCALE_OFFSET       1
#define TIMER_AUTORELOAD_OFFSET     2
#define TIMER_INTERRUPTS_OFFSET     3
#define TIMER_STATUS_OFFSET         4
#define TIMER_COUNTER_OFFSET        5

#define SERIAL_CONTROL_OFFSET       0
#define SERIAL_CONFIG_OFFSET        1
#define SERIAL_INTERRUPTS_OFFSET    2
#define SERIAL_STATUS_OFFSET        3
#define SERIAL_RXBUFFER_OFFSET      4
#define SERIAL_TXBUFFER_OFFSET      5

#define GPIO_MAX_PORT_NUM 22
#define GPIO_NDX_0      0
#define GPIO_NDX_1      1
#define GPIO_NDX_2      2
#define GPIO_NDX_3      3
#define GPIO_NDX_4      4
#define GPIO_NDX_5      5
#define GPIO_NDX_6      6
#define GPIO_NDX_7      7
#define GPIO_NDX_8      8
#define GPIO_NDX_9      9
#define GPIO_NDX_10     10
#define GPIO_NDX_11     11
#define GPIO_NDX_12     12
#define GPIO_NDX_13     13
#define GPIO_NDX_14     14
#define GPIO_NDX_15     15
#define GPIO_NDX_16     16
#define GPIO_NDX_17     17
#define GPIO_NDX_18     18
#define GPIO_NDX_19     19
#define GPIO_NDX_20     20
#define GPIO_NDX_21     21
#define GPIO_NDX_22     22
#define GPIO_NDX_23     23
#define GPIO_NDX_24     24
#define GPIO_NDX_25     25
#define GPIO_NDX_26     26
#define GPIO_NDX_27     27
#define GPIO_NDX_28     28
#define GPIO_NDX_29     29
#define GPIO_NDX_30     30
#define GPIO_NDX_31     31
#define GPIO_POS_0      0
#define GPIO_POS_1      1
#define GPIO_POS_2      2
#define GPIO_POS_3      3
#define GPIO_POS_4      4
#define GPIO_POS_5      5
#define GPIO_POS_6      6
#define GPIO_POS_7      7
#define GPIO_POS_8      8
#define GPIO_POS_9      9
#define GPIO_POS_10     10
#define GPIO_POS_11     11
#define GPIO_POS_12     12
#define GPIO_POS_13     13
#define GPIO_POS_14     14
#define GPIO_POS_15     15
#define GPIO_POS_16     16
#define GPIO_POS_17     17
#define GPIO_POS_18     18
#define GPIO_POS_19     19
#define GPIO_POS_20     20
#define GPIO_POS_21     21
#define GPIO_POS_22     22
#define GPIO_POS_23     23
#define GPIO_POS_24     24
#define GPIO_POS_25     25
#define GPIO_POS_26     26
#define GPIO_POS_27     27
#define GPIO_POS_28     28
#define GPIO_POS_29     29
#define GPIO_POS_30     30
#define GPIO_POS_31     31
#define GPIO_MSK_0      (1 << GPIO_POS_0)
#define GPIO_0          (GPIO_MSK_0     )
#define GPIO_MSK_1      (1 << GPIO_POS_1)
#define GPIO_1          (GPIO_MSK_1     )
#define GPIO_MSK_2      (1 << GPIO_POS_2)
#define GPIO_2          (GPIO_MSK_2     )
#define GPIO_MSK_3      (1 << GPIO_POS_3)
#define GPIO_3          (GPIO_MSK_3     )
#define GPIO_MSK_4      (1 << GPIO_POS_4)
#define GPIO_4          (GPIO_MSK_4     )
#define GPIO_MSK_5      (1 << GPIO_POS_5)
#define GPIO_5          (GPIO_MSK_5     )
#define GPIO_MSK_6      (1 << GPIO_POS_6)
#define GPIO_6          (GPIO_MSK_6     )
#define GPIO_MSK_7      (1 << GPIO_POS_7)
#define GPIO_7          (GPIO_MSK_7     )
#define GPIO_MSK_8      (1 << GPIO_POS_8)
#define GPIO_8          (GPIO_MSK_8     )
#define GPIO_MSK_9      (1 << GPIO_POS_9)
#define GPIO_9          (GPIO_MSK_9     )
#define GPIO_MSK_10     (1 << GPIO_POS_10)
#define GPIO_10         (GPIO_MSK_10    )
#define GPIO_MSK_11     (1 << GPIO_POS_11)
#define GPIO_11         (GPIO_MSK_11    )
#define GPIO_MSK_12     (1 << GPIO_POS_12)
#define GPIO_12         (GPIO_MSK_12    )
#define GPIO_MSK_13     (1 << GPIO_POS_13)
#define GPIO_13         (GPIO_MSK_13    )
#define GPIO_MSK_14     (1 << GPIO_POS_14)
#define GPIO_14         (GPIO_MSK_14    )
#define GPIO_MSK_15     (1 << GPIO_POS_15)
#define GPIO_15         (GPIO_MSK_15    )
#define GPIO_MSK_16     (1 << GPIO_POS_16)
#define GPIO_16         (GPIO_MSK_16    )
#define GPIO_MSK_17     (1 << GPIO_POS_17)
#define GPIO_17         (GPIO_MSK_17    )
#define GPIO_MSK_18     (1 << GPIO_POS_18)
#define GPIO_18         (GPIO_MSK_18    )
#define GPIO_MSK_19     (1 << GPIO_POS_19)
#define GPIO_19         (GPIO_MSK_19    )
#define GPIO_MSK_20     (1 << GPIO_POS_20)
#define GPIO_20         (GPIO_MSK_20    )
#define GPIO_MSK_21     (1 << GPIO_POS_21)
#define GPIO_21         (GPIO_MSK_21    )
#define GPIO_MSK_22     (1 << GPIO_POS_22)
#define GPIO_22         (GPIO_MSK_22    )
#define GPIO_MSK_23     (1 << GPIO_POS_23)
#define GPIO_23         (GPIO_MSK_23    )
#define GPIO_MSK_24     (1 << GPIO_POS_24)
#define GPIO_24         (GPIO_MSK_24    )
#define GPIO_MSK_25     (1 << GPIO_POS_25)
#define GPIO_25         (GPIO_MSK_25    )
#define GPIO_MSK_26     (1 << GPIO_POS_26)
#define GPIO_26         (GPIO_MSK_26    )
#define GPIO_MSK_27     (1 << GPIO_POS_27)
#define GPIO_27         (GPIO_MSK_27    )
#define GPIO_MSK_28     (1 << GPIO_POS_28)
#define GPIO_28         (GPIO_MSK_28    )
#define GPIO_MSK_29     (1 << GPIO_POS_29)
#define GPIO_29         (GPIO_MSK_29    )
#define GPIO_MSK_30     (1 << GPIO_POS_30)
#define GPIO_30         (GPIO_MSK_30    )
#define GPIO_MSK_31     (1 << GPIO_POS_31)
#define GPIO_31         (GPIO_MSK_31    )

#define TIMER_CONTROL_ENABLE_POS           4
#define TIMER_CONTROL_SINGLE_CYCLE_POS     5
#define TIMER_CONTROL_ARR_ENABLE_POS       6
#define TIMER_CONTROL_CLOCK_SOURCE_POS     7
#define TIMER_CONTROL_DIRECTION_POS       10
#define TIMER_CONTROL_ENABLE_MSK          (  1 << TIMER_CONTROL_ENABLE_POS       )
#define TIMER_CONTROL_SINGLE_CYCLE_MSK    (  1 << TIMER_CONTROL_SINGLE_CYCLE_POS)
#define TIMER_CONTROL_SINGLE_CYCLE        TIMER_CONTROL_SINGLE_CYCLE_MSK
#define TIMER_CONTROL_ARR_ENABLE_MSK      (  1 << TIMER_CONTROL_ARR_ENABLE_POS   )
#define TIMER_CONTROL_ARR_ENABLE          TIMER_CONTROL_ARR_ENABLE_MSK
#define TIMER_CONTROL_ENABLE              TIMER_CONTROL_ENABLE_MSK
#define TIMER_CONTROL_CLOCK_SOURCE_MSK    (0x7 << TIMER_CONTROL_CLOCK_SOURCE_POS)
#define TIMER_CONTROL_CLOCK_SOURCE_MAIN   (0x0 << TIMER_CONTROL_CLOCK_SOURCE_POS)
#define TIMER_CONTROL_CLOCK_SOURCE_PLL    (0x1 << TIMER_CONTROL_CLOCK_SOURCE_POS)
#define TIMER_CONTROL_CLOCK_SOURCE_EXTERN (0x2 << TIMER_CONTROL_CLOCK_SOURCE_POS)
#define TIMER_CONTROL_CLOCK_SOURCE_WWV    (0x3 << TIMER_CONTROL_CLOCK_SOURCE_POS)
#define TIMER_CONTROL_CLOCK_SOURCE_GPS    (0x4 << TIMER_CONTROL_CLOCK_SOURCE_POS)
#define TIMER_CONTROL_CLOCK_SOURCE_CESIUM (0x5 << TIMER_CONTROL_CLOCK_SOURCE_POS)
#define TIMER_CONTROL_DIRECTION_MSK       (  1 << TIMER_CONTROL_DIRECTION_POS)
#define TIMER_CONTROL_DIRECTION           TIMER_CONTROL_DIRECTION_MSK

#define TIMER_MAX_PRESCALER_LOW_VAL        0x7f
#define TIMER_PRESCALE_PRESCALER_LOW_POS   0
#define TIMER_PRESCALE_PRESCALER_LOW_MSK   (0x7f << TIMER_PRESCALE_PRESCALER_LOW_POS)
#define TIMER_MAX_PRESCALER_HIGH_VAL       0x1f
#define TIMER_PRESCALE_PRESCALER_HIGH_POS  25
#define TIMER_PRESCALE_PRESCALER_HIGH_MSK  (0x1f << TIMER_PRESCALE_PRESCALER_HIGH_POS)

#define TIMER_MAX_AUTORELOAD_VAL            0x00ffffff
#define TIMER_AUTORELOAD_AUTORELOAD_POS       0
#define TIMER_AUTORELOAD_AUTORELOAD_MSK     (0xffffff << TIMER_AUTORELOAD_AUTORELOAD_POS)


#define TIMER_INTERRUPTS_TIMEOUT_POS         8
#define TIMER_INTERRUPTS_UNDERFLOW_POS       9
#define TIMER_INTERRUPTS_TIMEOUT_MSK         (1 << TIMER_INTERRUPTS_TIMEOUT_POS)
#define TIMER_INTERRUPTS_TIMEOUT             TIMER_INTERRUPTS_TIMEOUT_MSK
#define TIMER_INTERRUPTS_UNDERFLOW_MSK      (1 << TIMER_INTERRUPTS_UNDERFLOW_POS)
#define TIMER_INTERRUPTS_UNDERFLOW           TIMER_INTERRUPTS_UNDERFLOW_MSK

#define TIMER_STATUS_TIMEOUT_POS            16
#define TIMER_STATUS_UNDERFLOW_POS          17
#define TIMER_STATUS_TIMEOUT_MSK            (1 << TIMER_STATUS_TIMEOUT_POS)
#define TIMER_STATUS_TIMEOUT                TIMER_STATUS_TIMEOUT_MSK
#define TIMER_STATUS_UNDERFLOW_MSK          (1 << TIMER_STATUS_UNDERFLOW_POS)
#define TIMER_STATUS_UNDERFLOW              TIMER_STATUS_UNDERFLOW_MSK

#define TIMER_COUNTER_COUNT_POS      8
#define TIMER_COUNTER_COUNT_MSK     (0xffffff << TIMER_COUNTER_COUNT_POS)


#define SERIAL_MAX_PORT_NUM              31

#define SERIAL_CONTROL_ENABLE_POS         0
#define SERIAL_CONTROL_RECV_POS           1
#define SERIAL_CONTROL_XMIT_POS           2
#define SERIAL_CONTROL_ENABLE_MSK        (  1  << SERIAL_CONTROL_ENABLE_POS)
#define SERIAL_CONTROL_ENABLE           SERIAL_CONTROL_ENABLE_MSK
#define SERIAL_CONTROL_RECV_MSK          (  1  << SERIAL_CONTROL_RECV_POS)
#define SERIAL_CONTROL_RECV         SERIAL_CONTROL_RECV_MSK
#define SERIAL_CONTROL_XMIT_MSK          (  1  << SERIAL_CONTROL_XMIT_POS)
#define SERIAL_CONTROL_XMIT         SERIAL_CONTROL_XMIT_MSK

#define SERIAL_CONFIG_MODE_POS            0
#define SERIAL_CONFIG_DATALEN_POS         1
#define SERIAL_CONFIG_PARITY_POS          3
#define SERIAL_CONFIG_ENDIAN_POS          5
#define SERIAL_CONFIG_POLARITY_POS        6
#define SERIAL_CONFIG_RXPORT_POS          7
#define SERIAL_CONFIG_TXPORT_POS         12
#define SERIAL_CONFIG_MODE_MSK           (  1  << SERIAL_CONFIG_MODE_POS)
#define SERIAL_CONFIG_MODE              SERIAL_CONFIG_MODE_MSK
#define SERIAL_CONFIG_DATALEN_MSK        (0x3  << SERIAL_CONFIG_DATALEN_POS)
#define SERIAL_CONFIG_DATALEN_8_BITS     (0x0  << SERIAL_CONFIG_DATALEN_POS)
#define SERIAL_CONFIG_DATALEN_16_BITS    (0x1  << SERIAL_CONFIG_DATALEN_POS)
#define SERIAL_CONFIG_DATALEN_32_BITS    (0x2  << SERIAL_CONFIG_DATALEN_POS)
#define SERIAL_CONFIG_PARITY_MSK         (0x3  << SERIAL_CONFIG_PARITY_POS)
#define SERIAL_CONFIG_PARITY_ODD         (0x0  << SERIAL_CONFIG_PARITY_POS)
#define SERIAL_CONFIG_PARITY_EVEN        (0x1  << SERIAL_CONFIG_PARITY_POS)
#define SERIAL_CONFIG_PARITY_CRC         (0x2  << SERIAL_CONFIG_PARITY_POS)
#define SERIAL_CONFIG_ENDIAN_MSK         (  1  << SERIAL_CONFIG_ENDIAN_POS)
#define SERIAL_CONFIG_ENDIAN            SERIAL_CONFIG_ENDIAN_MSK
#define SERIAL_CONFIG_POLARITY_MSK       (  1  << SERIAL_CONFIG_POLARITY_POS)
#define SERIAL_CONFIG_POLARITY          SERIAL_CONFIG_POLARITY_MSK
#define SERIAL_CONFIG_RXPORT_MSK         (0x1f << SERIAL_CONFIG_RXPORT_POS)
#define SERIAL_CONFIG_TXPORT_MSK         (0x1f << SERIAL_CONFIG_TXPORT_POS)


#define SERIAL_INTERRUPTS_RXREADY_POS       13
#define SERIAL_INTERRUPTS_TXREADY_POS       14
#define SERIAL_INTERRUPTS_RXOVERFLOW_POS    15
#define SERIAL_INTERRUPTS_TXUNDERFLOW_POS   16
#define SERIAL_INTERRUPTS_RXREADY_MSK       ((1 << SERIAL_INTERRUPTS_RXREADY_POS)
#define SERIAL_INTERRUPTS_RXREADY           SERIAL_INTERRUPTS_RXREADY_MSK
#define SERIAL_INTERRUPTS_TXREADY_MSK       ((1 << SERIAL_INTERRUPTS_TXREADY_POS)
#define SERIAL_INTERRUPTS_TXREADY           SERIAL_INTERRUPTS_TXREADY_MSK
#define SERIAL_INTERRUPTS_RXOVERFLOW_MSK ((1 << SERIAL_INTERRUPTS_RXOVERFLOW_POS)
#define SERIAL_INTERRUPTS_RXOVERFLOW        SERIAL_INTERRUPTS_RXOVERFLOW_MSK
#define SERIAL_INTERRUPTS_TXUNDERFLOW_MSK   ((1 << SERIAL_INTERRUPTS_TXUNDERFLOW_POS)
#define SERIAL_INTERRUPTS_TXUNDERFLOW       SERIAL_INTERRUPTS_TXUNDERFLOW_MSK

#define SERIAL_STATUS_RXREADY_POS        3
#define SERIAL_STATUS_TXREADY_POS        4
#define SERIAL_STATUS_RXOVERFLOW_POS     5
#define SERIAL_STATUS_TXUNDERFLOW_POS    6
#define SERIAL_STATUS_RXREADY_MSK       (1 << SERIAL_STATUS_RXREADY_POS)
#define SERIAL_STATUS_RXREADY           SERIAL_STATUS_RXREADY_MSK
#define SERIAL_STATUS_TXREADY_MSK       (1 << SERIAL_STATUS_TXREADY_POS)
#define SERIAL_STATUS_TXREADY           SERIAL_STATUS_TXREADY_MSK
#define SERIAL_STATUS_RXOVERFLOW_MSK    (1 << SERIAL_STATUS_RXOVERFLOW_POS)
#define SERIAL_STATUS_RXOVERFLOW        SERIAL_STATUS_RXOVERFLOW_MSK
#define SERIAL_STATUS_TXUNDERFLOW_MSK   (1 << SERIAL_STATUS_TXUNDERFLOW_POS)
#define SERIAL_STATUS_TXUNDERFLOW       SERIAL_STATUS_TXUNDERFLOW_MSK

#define SERIAL_RXBUFFER_DATA_POS     0
#define SERIAL_RXBUFFER_DATA_MSK    (0xffffffff << SERIAL_RXBUFFER_DATA_POS)

#define SERIAL_TXBUFFER_DATA_POS     0
#define SERIAL_TXBUFFER_DATA_MSK    (0xffffffff << SERIAL_TXBUFFER_DATA_POS)

#endif // #ifdef MCU_RAW_H
